hpx/lcos/detail/promise_lco.hpp

```diff
@@ -0,0 +1,217 @@
+//  Copyright (c) 2016      Thomas Heller
+//  Copyright (c) 2011      Bryce Adelstein-Lelbach
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef HPX_LCOS_DETAIL_PROMISE_LCO_HPP
+#define HPX_LCOS_DETAIL_PROMISE_LCO_HPP
+
+#include <hpx/lcos/base_lco_with_value.hpp>
+#include <hpx/lcos/detail/future_data.hpp>
+#include <hpx/runtime/components/component_type.hpp>
+#include <hpx/runtime/components/server/managed_component_base.hpp>
+#include <hpx/runtime/naming/id_type.hpp>
+#include <hpx/traits/component_type_database.hpp>
+#include <hpx/util/assert.hpp>
+#include <hpx/util/atomic_count.hpp>
+#include <hpx/util/unused.hpp>
+
+#include <boost/exception_ptr.hpp>
+#include <boost/intrusive_ptr.hpp>
+
+#include <utility>
+
+///////////////////////////////////////////////////////////////////////////////
+namespace hpx {
+namespace lcos {
+    namespace detail {
+        template <typename Result, typename RemoteResult>
+        class promise_lco;
+    }
+}
+}
+
+///////////////////////////////////////////////////////////////////////////////
+namespace hpx {
+namespace traits {
+    template <typename Result, typename RemoteResult>
+    struct managed_component_dtor_policy<
+        lcos::detail::promise_lco<Result, RemoteResult>>
+    {
+        typedef managed_object_is_lifetime_controlled type;
+    };
+}
+}
+
+///////////////////////////////////////////////////////////////////////////////
+namespace hpx {
+namespace lcos {
+    namespace detail {
+        template <typename Result, typename RemoteResult>
+        class promise_lco_base
+            : public lcos::base_lco_with_value<Result, RemoteResult>
+        {
+        protected:
+            typedef lcos::detail::future_data<Result>       shared_state_type;
+            typedef boost::intrusive_ptr<shared_state_type> shared_state_ptr;
+
+            typedef lcos::base_lco_with_value<Result, RemoteResult> base_type;
+
+            typedef typename base_type::result_type result_type;
+
+        public:
+            promise_lco_base(shared_state_ptr const& shared_state)
+              : shared_state_(shared_state)
+            {
+            }
+
+            void set_value(RemoteResult&& result)
+            {
+                HPX_ASSERT(shared_state_);
+                shared_state_->set_data(std::move(result));
+            }
+
+            void set_exception(boost::exception_ptr const& e)
+            {
+                HPX_ASSERT(shared_state_);
+                shared_state_->set_exception(e);
+            }
+
+            // This is the component id. Every component needs to have an
+            // embedded
+            // enumerator 'value' which is used by the generic action
+            // implementation
+            // to associate this component with a given action.
+            enum
+            {
+                value = components::component_promise
```

Comment:
This loses the distinction of different promise types in terms of their component type.